### PR TITLE
Q scale factor from preferences

### PR DESCRIPTION
**Pull request by daschuer**
This allows to set the environment variable QT_SCALE_FACTOR from preferences. 

Unfortunately it is required to parse mixxx.cfg with custom code before initializing the QApp.
A workaround would be to use a private QT API for HID scaling which is also not nice. 

This allows also to scale Mixxx by 50 % on a 200 % System while avoiding to go below 100 % for the overall scaling.

### CODE REVIEW

**Comment by daschuer** [comment]
This restores also a Sandbox::checkSandboxed(); call.
Was it lost or is it not required in 2.4? 

This should be clarified before mergeing this.

File: src/main.cpp [hunk 1]
```diff
@@ -26,6 +28,9 @@ namespace {
 constexpr int kFatalErrorOnStartupExitCode = 1;
 constexpr int kParseCmdlineArgsErrorExitCode = 2;
 
+constexpr char kScaleFactorEnvVar[] = "QT_SCALE_FACTOR";
+const QString kStrScaleFactor = QStringLiteral("ScaleFactor");
```
Comment (uklotzde) [inline src/main.cpp:32]: Please don't introduce custom type prefixes. Instead name the constant according to its purpose, i.e. a config key as far as I understood.

**Review by uklotzde — CHANGES_REQUESTED** [review]

File: src/preferences/dialog/dlgprefinterface.cpp [hunk 2]
```diff
@@ -42,6 +43,16 @@ DlgPrefInterface::DlgPrefInterface(
     // get the pixel ratio to display a crisp skin preview when Mixxx is scaled
     m_dDevicePixelRatio = getDevicePixelRatioF(this);
 
+    // Calculate the minimum scale factor that leads to a device picel ratio of 1.0
```
Comment (ronso0) [inline src/preferences/dialog/dlgprefinterface.cpp:46]:
```suggestion
    // Calculate the minimum scale factor that leads to a device pixel ratio of 1.0
```

File: src/preferences/dialog/dlgprefinterface.cpp [hunk 3]
```diff
@@ -42,6 +43,16 @@ DlgPrefInterface::DlgPrefInterface(
...
+    // m_dDevicePixelRatio must not drop below 1.0 because this creats an
```
Comment (ronso0) [inline src/preferences/dialog/dlgprefinterface.cpp:47]:
```suggestion
    // m_dDevicePixelRatio must not drop below 1.0 because this creates an
```

**Review by ronso0 — COMMENTED** [review]
just some typos.
will test this soon.

**Comment by daschuer** [comment]
Done.

**Comment by Holzhaus** [comment]
What is the use case for using a scale factor different from the rest of the system?

**Comment by daschuer** [comment]
To bee able using the Deere skin on a full HD 15'' device. Windows scales such a device to 150% by default. That works nice for most Applications, but does not work for Mixxx. You can also consider the other way around. 
We have many users asking for the env variable trick ...

**Comment by ronso0** [comment]
> What is the use case for using a scale factor different from the rest of the system?

adjust the GUI to increased eye/screen distance when djing.
work setting, laptop on desk: ~0.5m
DJ setting, standing, laptop behind/next to controller: up to 1.0m

File: src/preferences/dialog/dlgprefinterface.cpp [hunk 4]
```diff
@@ -44,6 +43,16 @@ DlgPrefInterface::DlgPrefInterface(
     // get the pixel ratio to display a crisp skin preview when Mixxx is scaled
     m_dDevicePixelRatio = getDevicePixelRatioF(this);
 
+    // Calculate the minimum scale factor that leads to a device pixel ratio of 1.0
+    // m_dDevicePixelRatio must not drop below 1.0 because this creates an
+    // unusable GUI with visual artefacts
+    double initialScalefactor = CmdlineArgs::Instance().getScaleFactor();
```
Comment (uklotzde) [inline src/preferences/dialog/dlgprefinterface.cpp:49]: `initialScaleFactor`

File: src/preferences/dialog/dlgprefinterface.cpp [hunk 5]
```diff
@@ -248,13 +253,11 @@ void DlgPrefInterface::slotUpdate() {
     m_localeOnUpdate = m_pConfig->getValueString(ConfigKey("[Config]", "Locale"));
     ComboBoxLocale->setCurrentIndex(ComboBoxLocale->findData(m_localeOnUpdate));
 
-    checkBoxScaleFactorAuto->setChecked(m_pConfig->getValue(
-            ConfigKey("[Config]", "ScaleFactorAuto"), m_bUseAutoScaleFactor));
-
     // The spinbox shows a percentage but Mixxx stores a multiplication factor
     // with 1.00 as no scaling, so multiply the stored value by 100.
     spinBoxScaleFactor->setValue(m_pConfig->getValue(
```
Comment (uklotzde) [inline src/preferences/dialog/dlgprefinterface.cpp:258]: `kScaleFactorConfigKey`

File: src/preferences/dialog/dlgprefinterface.cpp [hunk 6]
```diff
@@ -396,14 +374,8 @@ void DlgPrefInterface::slotApply() {
...
-    m_pConfig->setValue(
-            ConfigKey("[Config]", "ScaleFactorAuto"), m_bUseAutoScaleFactor);
-    if (m_bUseAutoScaleFactor) {
-        m_pConfig->setValue(
-                ConfigKey("[Config]", "ScaleFactor"), m_dScaleFactorAuto);
-    } else {
-        m_pConfig->setValue(ConfigKey("[Config]", "ScaleFactor"), m_dScaleFactor);
-    }
+    double scaleFactor = spinBoxScaleFactor->value() / 100;
+    m_pConfig->setValue(ConfigKey("[Config]", "ScaleFactor"), scaleFactor);
```
Comment (uklotzde) [inline src/preferences/dialog/dlgprefinterface.cpp:378]: `kScaleFactorConfigKey`

**Comment by daschuer** [comment]
Done.

File: src/preferences/dialog/dlgprefinterface.cpp [hunk 7]
```diff
@@ -149,13 +171,12 @@ DlgPrefInterface::DlgPrefInterface(
...
-    labelScaleFactor->hide();
-
     // Start in fullscreen mode
-    checkBoxStartFullScreen->setChecked(m_pConfig->getValueString(
-                    ConfigKey("[Config]", "StartInFullscreen")).toInt()==1);
+    checkBoxStartFullScreen->setChecked(
+            m_pConfig
+                    ->getValueString(
+                            ConfigKey(kConfigGroup, kStartInFullscreenKey))
+                    .toInt() == 1);
```
Comment (ronso0) [inline src/preferences/dialog/dlgprefinterface.cpp:179]: there are some ugly linebreaks now.
what about this, precommit agrees
```suggestion
            m_pConfig->getValueString(
                             ConfigKey(kConfigGroup, kStartInFullscreenKey))
                    .toInt() == 1);
```

Comment (Holzhaus) [reply src/preferences/dialog/dlgprefinterface.cpp:179]: Why not make the whole config key a constant? Then the line would be shorter.

**Comment by daschuer** [comment]
Line breaks improved

**Comment by daschuer** [comment]
@uklotzde: merge?

**Review by uklotzde — APPROVED** [review]
Please resolve the conflicts.

**Comment by daschuer** [comment]
Done

File: src/main.cpp [hunk 8]
```diff
@@ -48,6 +53,41 @@ int runMixxx(MixxxApplication* pApp, const CmdlineArgs& args) {
...
+    // We cannot use ConfigObject, because it depends on MixxxApplication
+    // but the scale factor is read during it's constructor.
+    // QHighDpiScaling can not be used afterwards because it is private.
+    auto cfgFile = QFile(QDir(pArgs->getSettingsPath()).filePath(MIXXX_SETTINGS_FILE));
+    if (cfgFile.open(QFile::ReadOnly | QFile::Text)) {
+        QTextStream in(&cfgFile);
+        QString strScaleFactor;
+        QString line = in.readLine();
+        while (!line.isNull()) {
+            if (line.startsWith(kScaleFactorConfigKey)) {
```
Comment (Holzhaus) [inline src/main.cpp:76]: Have you considered using `QSettings` instead of hacking a new config parser here?

Comment (daschuer) [reply src/main.cpp:76]: Yes, unfortunately the attempt failed because QSettings expects a "=" between key and value, which is not used in our ini file.

Comment (Holzhaus) [reply src/main.cpp:76]: Ok. It looks like only the constructor depends on QApplication to construct file paths. Can you just add a ConfigObject static method like `ConfigObject::fromFilePath(path)`?

File: src/main.cpp [hunk 9]
```diff
@@ -1,11 +1,13 @@
 #include <QApplication>
 #include <QDir>
+#include <QSettings>
```
Comment (uklotzde) [inline src/main.cpp:3]: QSettings seems to be unused and a left-over from previous experiments?

File: src/main.cpp [hunk 10]
```diff
@@ -48,6 +53,41 @@ int runMixxx(MixxxApplication* pApp, const CmdlineArgs& args) {
...
+        }
+    }
+    // We cannot use ConfigObject, because it depends on MixxxApplication
+    // but the scale factor is read during it's constructor.
+    // QHighDpiScaling can not be used afterwards because it is private.
+    auto cfgFile = QFile(QDir(pArgs->getSettingsPath()).filePath(MIXXX_SETTINGS_FILE));
+    if (cfgFile.open(QFile::ReadOnly | QFile::Text)) {
+        QTextStream in(&cfgFile);
+        QString strScaleFactor;
+        QString line = in.readLine();
```
Comment (uklotzde) [inline src/main.cpp:74]: Better use `in.readLine().trimmed()` (here an below) in case users edited their config file manually.

**Review by uklotzde — CHANGES_REQUESTED** [review]

**Comment by daschuer** [comment]
Done. Now it uses the ConfigObject parser.

**Comment by daschuer** [comment]
@uklotzde: Please have another look.

File: src/util/cmdlineargs.h [hunk 11]
```diff
@@ -44,6 +44,13 @@ class CmdlineArgs final {
     const QString& getResourcePath() const { return m_resourcePath; }
     const QString& getTimelinePath() const { return m_timelinePath; }
 
+    void storeScaleFactor(double scaleFactor) {
```
Comment (uklotzde) [inline src/util/cmdlineargs.h:47]: Why "store"? This is just a simple setter. There is no side-effect.

Comment (daschuer) [reply src/util/cmdlineargs.h]: renamed to storeScaleFactor()

**Review by uklotzde — APPROVED** [review]
Time will tell if this fixes the scaling issues that have been reported frequently. Thank you! LGTM

**Pull request merged**